### PR TITLE
chore: create inline UIView that checks if there are inline messages in local queue

### PR DESCRIPTION
**Pull request by levibostian**
Part of: https://linear.app/customerio/issue/MBL-310/create-a-uikit-uiview-that-displays-an-inline-in-app-message-sent-from

Create a new public UIView that customers can add to their app's UI to display inline Views. This commit does not display messages yet. It only checks if there is an in-app message available to display inline with the equal elementId.

Testing:
Automated tests added in the commit. Some boilerplate code was added to allow mocking some dependencies and testing functions that were previously not very testable.

Reviewer notes:
This commit only handles the use case of inline Views being able to display inline messages after a fetch has already completed in the SDK. Future changes will handle when an inline View is visible and a fetch completes.

---

**Stack**:
- #720
- #718
- #717
- #716 ⬅

⚠️ *Part of a stack created by [spr](https://github.com/ejoffe/spr). Do not merge manually using the UI - doing so may have unexpected results.*

### CODE REVIEW

File: Sources/MessagingInApp/Gist/Gist.swift
Comment (levibostian) [inline Sources/MessagingInApp/Gist/Gist.swift:163]: Only logic added for inline. Rest of code modified is to make the code testable.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [inline Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:85]: Returns all inline messages with a matching element id.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [inline Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:167]: Modified logic here to add all inline messages to local queue.

Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:167]: // TODO: I need to verify that when page rule triggers and SDK checks if we need to display any modal messages it filters out inline messages.

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (levibostian) [inline Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: Added this comment because[ I asked in-app squad about it](https://customerio.slack.com/archives/C036FTE6UJ0/p1715967637970009). I wasn't sure if this impacts the messages we return in queuemanager.getInlineMessages and it doesn't.

File: Tests/MessagingInApp/Extensions/GistExtensions.swift
Comment (levibostian) [inline Tests/MessagingInApp/Extensions/GistExtensions.swift:1]: Boilerplate for testing. Allows us to create test instances of Message for inline tests

File: Tests/Shared/NSCoder.swift
Comment (levibostian) [inline Tests/Shared/NSCoder.swift:1]: Required to add this for `InAppMessageViewTest` to create instances of `InAppMessageView`.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [inline Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:89]: I moved this block into a class function so I could write automated tests against the logic.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (Ahmed-Ali) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:85]: what is the use case where we can have multiple messages matching the same element id?

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: based on the comment beside it, shouldn't this be a message id? the comment says it identifies the message but the property name is queue id

File: Sources/MessagingInApp/Views/InAppMessageView.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Views/InAppMessageView.swift:42]: 👍

File: Sources/MessagingInApp/Views/InAppMessageView.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Views/InAppMessageView.swift:65]: delete

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:22]: public mutable property is almost always not good choice.
let alone the no-threading-protection

**Review by Ahmed-Ali — CHANGES_REQUESTED** [review]
requesting changes mainly for the exposed mutable var in the message queue maanger

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Shahroz16) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: I think the comment is a bit confusing, it probably uniquely identifies an in-app message in the queue for this user. 

Message-id is the unique identifier for the in-app message.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:85]: [Gist's backend service can return multiple in-app messages with the same element id. You have to use the priority property on the message](https://customerio.slack.com/archives/C036FTE6UJ0/p1716306131792019?thread_ts=1716304555.465419&cid=C036FTE6UJ0). 

I will modify this function in this PR to sort the array by priority before returning.

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: [I was told that the `queueId` is a unique identifier for a message](https://customerio.slack.com/archives/C036FTE6UJ0/p1715967782604259?thread_ts=1715967637.970009&cid=C036FTE6UJ0). 

[The `messageId` refers to templates and so it may not be unique](https://customerio.slack.com/archives/C036FTE6UJ0/p1716291837428599?thread_ts=1715967637.970009&cid=C036FTE6UJ0).

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:29]: I needed to generate a mock for testing. We generate mocks by writing protocols. You can't generate a protocol with default parameters. Instead, you must write all versions of the functions you plan to provide. So, we need 2 `setup()` functions to replace the 1 `setup()` function that used default parameter values.

File: Sources/MessagingInApp/Gist/Gist.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Gist.swift:125]: The comment might be adding confusion. It's not meant to say that we will never have inline messages run the logic in this function, but instead do not call it at this time. 

To make sure that `shownModalMessageQueueIds` does not get modified for inline messages, I added the `guard`. 

In milestone 1, the only way for a message to no longer be shown is have it expire. So, removing from the local store and modifying `shownModalMessageQueueIds` at this time is not wanted.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:22]: The property is `internal` as that's the swift default. It's made internal so we can reference it in our automated tests. 

As far as no-threading, that's a good call-out. I can make it have safe reading and writing access as we do in the codebase today.

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift]: [List is sorted by priority before being returned](https://github.com/customerio/customerio-ios/pull/716/files#diff-ec01291f32e74af14510929ca3d297f876bb1882bd4fdd5008cadca4dff331bcR73).

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift]: I [updated](https://github.com/customerio/customerio-ios/pull/716/files#diff-ec01291f32e74af14510929ca3d297f876bb1882bd4fdd5008cadca4dff331bcR19-R23) the `localMessageStore` and `interval` to have atomic read and write access.

File: Sources/MessagingInApp/Gist/Gist.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Gist.swift]: [I updated the comment to try and be more clear](https://github.com/customerio/customerio-ios/pull/716/files#diff-05212de0e1f067b635cf22ad2778b1d51110ddeb8d5615b893ea42b4324e9ca0R125-R127). 

The biggest thing I am trying to do is protecting the `shownModalMessageQueueIds` from being inserted with inline messages.

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Shahroz16) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: I agree `messageId` refers to template, so we can have multiple messages with the same message id in the queue. But all of them will have a unique `queueId` in that queue and the same message when expired and added again will have a new `queueId` so instead of `backend system` I would rather put it as `queue`

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (Ahmed-Ali) [reply Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:31]: I would agree that it is better to keep one version with no default value. Default value can make it harder to understand the default behavior, or even worse, change the default behavior without introducing API change

File: Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Gist/Managers/MessageQueueManager.swift:131]: Let's make sure we don't merge changes with commented out code please

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Ahmed-Ali) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: I still find the name `queueId` confusing. It makes me think that there are multiple queues, and here we are referencing one specific queue by its id.

The `messageId` referencing a template is also confusing but I would be okay that it is out of scope for this PR.

For the property referenced above we can come up with something better, maybe something like `messageInstanceId` or anything else that associates the uniqueness and id directly to the message not the queue

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Gist/Managers/Models/Message.swift:23]: if it is not big change, let's rename this to `templateId` as a messageId referencing a template isn't very conveying to what it actually is.

File: Sources/MessagingInApp/Views/InAppMessageView.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Views/InAppMessageView.swift:50]: why an empty method?
if it will be implemented later, then let's add it later.

File: Sources/MessagingInApp/Views/InAppMessageView.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Views/InAppMessageView.swift:65]: left over bracket, please clean it up

File: Sources/MessagingInApp/Views/InAppMessageView.swift
Comment (Ahmed-Ali) [inline Sources/MessagingInApp/Views/InAppMessageView.swift:68]: another empty private method, can be added when it is implemented

**Review by Ahmed-Ali — CHANGES_REQUESTED** [review]

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (Ahmed-Ali) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:23]: I am fine with not doing this change in this PR for sure. Let's just not lose track of it as these hard to reason about names can easily fire back later.

**Review by Ahmed-Ali — APPROVED** [review]

File: Sources/MessagingInApp/Gist/Managers/Models/Message.swift
Comment (levibostian) [reply Sources/MessagingInApp/Gist/Managers/Models/Message.swift:21]: Opened a new PR to do the renaming instead of a TODO. https://github.com/customerio/customerio-ios/pull/730

**Pull request merged**
